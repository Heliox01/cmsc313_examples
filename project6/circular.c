#include "circular.h"

/*
 * This is where you should write all your code (except any additional
 * tests you choose to include).
 */


/*
 * First define what the real _cbuf struct should look like. Remember,
 * this is opaque to the users of your library so you can do whatever
 * you like in here. The _cbuf struct is typedefd in the header to the
 * cbuf type so you're free to use either type mostly
 * interchangably. The one exception is that sizeof calls must use the
 * _cbuf type.
 */

struct _cbuf {
  // whatever you want can go in here
};

/*
 * Now define your implementations of all of the functions given in
 * the header. The test programs will not link successfully until each
 * function they use is defined here.
 */
