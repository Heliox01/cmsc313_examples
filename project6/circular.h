#ifndef CIRCULAR_H
#define CIRCULAR_H

/* provides size_t */
#include <stddef.h>

/*
 * This is an opaque data structure. You may define the _cbuf struct
 * inside your C file and include whatever fields you like. The user
 * of your library should never know or care about the internals of
 * your cbuf struct.
 */
struct _cbuf;
typedef struct _cbuf cbuf;

/*
 * This is the type that users will store inside your data
 * structure. Its definition is exposed to users because they will
 * need to be able to access its fields.
 *
 * Your implementation should not need to modify the provided
 * definition of this datastructure.
 */
typedef struct {
  double rate;
  unsigned int time;
} quote;


/*
 * These are the "public" functions used for interacting with your
 * datastructure. You are free to define more functions to be used as
 * internal implementation details but you should define them only in
 * your C file so that the users of your datastructure will not need
 * to know or care about them. If you wish, you can create a seperate
 * "circular_internal.h" header that has definitions of these internal
 * functions so that your tests can access them.
 */


/*
 * Initialize a circular buffer and return a pointer that can be used
 * to specify that circular buffer in future function calls.
 */
cbuf* cbuf_alloc();

/*
 * Deallocate all dynamically allocated memory associated with the
 * circular buffer referenced by cb_ptr. The programmer using your ADT
 * promises to never use the value in cb_ptr again.
 */
void cbuf_free(cbuf *cb_ptr);

/*
 * Add a new currency exchange quote to the specified circular
 * buffer. The time is in seconds since midnight. This function should
 * remove any quotes in the circular buffer that are more than 5
 * minutes older than this quote. You should assume that the
 * timestamps given to cbuf_update() are in non-decreasing
 * order. During a cbuf_update(), the memory for the circular buffer
 * may have to grow or shrink according to the rules given in the
 * project description.
 */
void cbuf_update(cbuf *cb_ptr, unsigned int time, double rate);

/*
 * Return the average exchange rate of the quotes currently stored in
 * the buffer.
 */
double cbuf_average(cbuf *cb_ptr);

/*
 * Return a pointer to the earliest quote currently stored in the
 * buffer. (Earliest = smallest timestamp) The programmer using your
 * ADT promises to just "look at" the quote and not alter it in any
 * way.
 */
const quote *cbuf_start(cbuf *cb_ptr);

/*
 * Return a pointer to the latest quote currently stored in the
 * buffer. (Latest = largest timestamp) The programmer using your ADT
 * promises to just "look at" the quote and not alter it in any way.
 */
const quote *cbuf_end(cbuf *cb_ptr);

/*
 * Return the number of elements currently stored in the circular
 * buffer.
 */
size_t cbuf_size(cbuf *cb_ptr);

/*
 * Return the current capacity of the circular buffer (will be greater
 * than or equal to the number of elements in the buffer). When
 * cbuf_size(x) == cbuf_capacity(x) then your circular buffer may need
 * to grow when the next element is inserted.
 */
size_t cbuf_capacity(cbuf *cb_ptr);

#endif
