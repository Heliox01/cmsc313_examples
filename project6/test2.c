/*
 * Test out circular buffer for foreign
 * exchange quotes.
 */

#include <stdio.h>
#include <stdlib.h>
#include "circular.h"

int main(int argc, char** argv) {
  cbuf *cb1 = cbuf_alloc();

  cbuf_update(cb1, 60, 1.291);
  cbuf_update(cb1, 63, 1.287);
  cbuf_update(cb1, 63, 1.231);
  cbuf_update(cb1, 69, 1.229);
  cbuf_update(cb1, 72, 1.247);


  cbuf_update(cb1, 361, 1.291);


  cbuf_update(cb1, 411, 1.291);
  cbuf_update(cb1, 412, 1.281);
  cbuf_update(cb1, 413, 1.292);
  cbuf_update(cb1, 414, 1.284);
  cbuf_update(cb1, 414, 1.290);

  cbuf_update(cb1, 511, 1.241);
  cbuf_update(cb1, 512, 1.251);
  cbuf_update(cb1, 513, 1.232);
  cbuf_update(cb1, 514, 1.202);
  cbuf_update(cb1, 517, 1.119);

  cbuf_update(cb1, 551, 1.080);
  cbuf_update(cb1, 552, 1.081);
  cbuf_update(cb1, 553, 1.079);
  cbuf_update(cb1, 554, 1.088);
  cbuf_update(cb1, 561, 1.072);
  cbuf_update(cb1, 562, 1.113);
  cbuf_update(cb1, 563, 1.091);
  cbuf_update(cb1, 564, 1.092);
  cbuf_update(cb1, 571, 1.089);
  cbuf_update(cb1, 572, 1.073);
  cbuf_update(cb1, 573, 1.061);
  cbuf_update(cb1, 574, 1.111);
  cbuf_update(cb1, 581, 1.119);
  cbuf_update(cb1, 582, 1.123);
  cbuf_update(cb1, 583, 1.151);
  cbuf_update(cb1, 584, 1.153);

  double avg;
  avg = cbuf_average(cb1);
  printf("Average rate = %f\n", avg);

  const quote *q_start, *q_end;
  q_start = cbuf_start(cb1);
  q_end = cbuf_end(cb1);

  printf("Start: time = %d, rate = %f\n", q_start->time, q_start->rate);
  printf("End:   time = %d, rate = %f\n", q_end->time, q_end->rate);

  cbuf_free(cb1);
  return 0;
}
