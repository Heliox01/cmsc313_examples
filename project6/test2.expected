Average rate = 1.161111
Start: time = 361, rate = 1.291000
End:   time = 584, rate = 1.153000
